### PR TITLE
[pipeline] Add chatglm pipeline 

### PR DESCRIPTION
**Pull request by CjhHa1**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`
#4286 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

this pr adds pipeline forwards for chatglm

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/shardformer/modeling/chatglm.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+                                                  dtype=inputs_embeds.dtype)
+            if attention_mask is not None:
+                attention_mask = torch.cat([attention_mask.new_ones((batch_size, self.pre_seq_len)), attention_mask],
+                                           dim=-1)
+
+        if full_attention_mask is None:
+            if (attention_mask is not None and not attention_mask.all()) or (past_key_values and seq_length != 1):
+                full_attention_mask = self.get_masks(input_ids, past_key_values, padding_mask=attention_mask)
+        # Rotary positional embeddings
+        rotary_pos_emb = self.rotary_pos_emb(self.seq_length)
```
Comment (Fridge003) [inline colossalai/shardformer/modeling/chatglm.py:80]: The part of rotary_pos_emb might be put under "if stage.is_first_stage()"

Comment (CjhHa1) [reply colossalai/shardformer/modeling/chatglm.py:80]: rotary pos embed is needed for every layer so it must be constructed in each stage

Comment (Fridge003) [reply colossalai/shardformer/modeling/chatglm.py:80]: Then rotary_pos_emb needs to be appended in held_layers for every stage.

Comment (CjhHa1) [reply colossalai/shardformer/modeling/chatglm.py:80]: done

**Review by Fridge003 — APPROVED** [review]

File: colossalai/shardformer/policies/chatglm.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+__all__ = ['ChatGLMPolicy', 'ChatGLMModelPolicy', 'ChatGLMForConditionalGenerationPolicy']
+
+
+class ChatGLMPolicy(Policy):
+
+    def config_sanity_check(self):
+        pass
+
+    def preprocess(self):
+        # Resize embedding
```
Comment (ver217) [inline colossalai/shardformer/policies/chatglm.py:30]: Check whether tp is enabled here.

File: tests/kit/model_zoo/transformers/chatglm.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    attention_mask = torch.tensor([[1, 1, 1, 1, 1, 1]])
+    return dict(input_ids=input_ids, attention_mask=attention_mask)
+
+
+# define output transform function
+output_transform_fn = lambda x: x
+
+# define loss function
+loss_fn_for_chatglm_model = lambda x: x.last_hidden_state.mean()
+loss_fn = lambda x: x.logits.mean()
```
Comment (ver217) [inline tests/kit/model_zoo/transformers/chatglm.py:25]: Use real loss here.

File: colossalai/shardformer/policies/chatglm.py (hunk 2 above)
Comment (CjhHa1) [reply colossalai/shardformer/policies/chatglm.py:30]: done

File: tests/test_shardformer/test_model/test_shard_chatglm.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        chatglm_model = org_model.transformer
+        shard_chatglm_model = sharded_model.transformer
+
+    # check attention grad
+    org_grad = chatglm_model.encoder.layers[0].self_attention.query_key_value.weight.grad
+    shard_grad = shard_chatglm_model.encoder.layers[0].self_attention.query_key_value.weight.grad
+    shard_weight = shard_chatglm_model.encoder.layers[0].self_attention.query_key_value.weight
+
+    if is_distributed_tensor(shard_weight) or is_customized_distributed_tensor(shard_weight):
+        shard_grad_list = [torch.zeros([*shard_grad.shape]).to('cuda') for _ in range(2)]
```
Comment (ver217) [inline tests/test_shardformer/test_model/test_shard_chatglm.py:49]: You can directly use `torch.zeros_like()`.

File: tests/kit/model_zoo/transformers/chatglm.py (hunk 3 above)
Comment (CjhHa1) [reply tests/kit/model_zoo/transformers/chatglm.py]: done

File: tests/test_shardformer/test_model/test_shard_chatglm.py (hunk 4 above)
Comment (CjhHa1) [reply tests/test_shardformer/test_model/test_shard_chatglm.py:49]: done
